File: solutions/java/src/main/java/com/epi/TeamPhoto1.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Arrays;

// @include
class Player implements Comparable<Player> {
  public Integer height;

  public Player(Integer h) { height = h; }

  @Override
  public int compareTo(Player that) {
    return height.compareTo(that.height);
  }
}

class Team {
  public Team(int[] height) {
    players = new Player[height.length];
    for (int i = 0; i < height.length; ++i) {
      players[i] = new Player(height[i]);
    }
  }

  // Checks if A can be placed in front of B.
  public static boolean validPlacementExists(Team A, Team B) {
    Player[] ASorted = A.sortPlayersByHeight();
    Player[] BSorted = B.sortPlayersByHeight();
    for (int i = 0; i < ASorted.length && i < BSorted.length; ++i) {
      if (ASorted[i].compareTo(BSorted[i]) >= 0) {
        return false;
      }
    }
    return true;
  }

  private Player[] sortPlayersByHeight() {
    Player[] sortedPlayers = players;
    Arrays.sort(sortedPlayers);
    return sortedPlayers;
  }

  private Player[] players;
}
// @exclude

class TeamPhoto1 {
  public static void main(String[] args) {
    int[] height = new int[] {1, 5, 4};
    Team t1 = new Team(height);
    height = new int[] {2, 3, 4};
    Team t2 = new Team(height);
    assert(!Team.validPlacementExists(t1, t2) &&
           !Team.validPlacementExists(t2, t1));
    height = new int[] {0, 3, 2};
    Team t3 = new Team(height);
    assert(
        Team.validPlacementExists(t3, t1) && !Team.validPlacementExists(t1, t3) &&
        Team.validPlacementExists(t3, t2) && !Team.validPlacementExists(t1, t2));
  }
}

File: solutions/cpp/Team_photo_1.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <vector>

using std::vector;

// @include
struct Player {
  bool operator<(const Player& that) const { return height < that.height; }

  int height;
};

class Team {
 public:
  explicit Team(const vector<int>& height) {
    for (const int& h : height) {
      members_.emplace_back(Player{h});
    }
  }

  bool operator<(const Team& that) const {
    vector<Player> this_sorted(SortHeightMembers());
    vector<Player> that_sorted(that.SortHeightMembers());
    for (int i = 0; i < this_sorted.size() && i < that_sorted.size(); ++i) {
      if (!(this_sorted[i] < that_sorted[i])) {
        return false;
      }
    }
    return true;
  }

 private:
  vector<Player> SortHeightMembers() const {
    vector<Player> sorted_members(members_);
    sort(sorted_members.begin(), sorted_members.end());
    return sorted_members;
  }

  vector<Player> members_;
};
// @exclude

int main(int argc, char* argv[]) {
  vector<int> height(3);
  height[0] = 1, height[1] = 5, height[2] = 4;
  Team t1(height);
  height[0] = 2, height[2] = 3, height[2] = 4;
  Team t2(height);
  assert(!(t1 < t2) && !(t2 < t1));
  height[0] = 0, height[1] = 3, height[2] = 2;
  Team t3(height);
  assert(t3 < t1 && !(t1 < t3) && t3 < t2 && !(t1 < t2));
  return 0;
}
